In [4]:
import pandas as pd 
import numpy as np
import matplotlib as plt
import seaborn as sns

In [12]:
df = pd.read_csv("client_data (1).csv")
df

id                     channel_sales  \
0      24011ae4ebbe3035111d65fa7c15bc57  foosdfpfkusacimwkcsosbicdxkicaua   
1      d29c2c54acc38ff3c0614d0a653813dd                           MISSING   
2      764c75f661154dac3a6c254cd082ea7d  foosdfpfkusacimwkcsosbicdxkicaua   
3      bba03439a292a1e166f80264c16191cb  lmkebamcaaclubfxadlmueccxoimlema   
4      149d57cf92fc41cf94415803a877cb4b                           MISSING   
...                                 ...                               ...   
14601  18463073fb097fc0ac5d3e040f356987  foosdfpfkusacimwkcsosbicdxkicaua   
14602  d0a6f71671571ed83b2645d23af6de00  foosdfpfkusacimwkcsosbicdxkicaua   
14603  10e6828ddd62cbcf687cb74928c4c2d2  foosdfpfkusacimwkcsosbicdxkicaua   
14604  1cf20fd6206d7678d5bcafd28c53b4db  foosdfpfkusacimwkcsosbicdxkicaua   
14605  563dde550fd624d7352f3de77c0cdfcd                           MISSING   

       cons_12m  cons_gas_12m  cons_last_month  date_activ    date_end  \
0             0         54946                0  2013-06-15  2016-06-15   
1          4660             0                0  2009-08-21  2016-08-30   
2           544             0                0  2010-04-16  2016-04-16   
3          1584             0                0  2010-03-30  2016-03-30   
4          4425             0              526  2010-01-13  2016-03-07   
...         ...           ...              ...         ...         ...   
14601     32270         47940                0  2012-05-24  2016-05-08   
14602      7223             0              181  2012-08-27  2016-08-27   
14603      1844             0              179  2012-02-08  2016-02-07   
14604       131             0                0  2012-08-30  2016-08-30   
14605      8730             0                0  2009-12-18  2016-12-17   

      date_modif_prod date_renewal  forecast_cons_12m  ...  has_gas  imp_cons  \
0          2015-11-01   2015-06-23               0.00  ...        t      0.00   
1          2009-08-21   2015-08-31             189.95  ...        f      0.00   
2          2010-04-16   2015-04-17              47.96  ...        f      0.00   
3          2010-03-30   2015-03-31             240.04  ...        f      0.00   
4          2010-01-13   2015-03-09             445.75  ...        f     52.32   
...               ...          ...                ...  ...      ...       ...   
14601      2015-05-08   2014-05-26            4648.01  ...        t      0.00   
14602      2012-08-27   2015-08-28             631.69  ...        f     15.94   
14603      2012-02-08   2015-02-09             190.39  ...        f     18.05   
14604      2012-08-30   2015-08-31              19.34  ...        f      0.00   
14605      2009-12-18   2015-12-21             762.41  ...        f      0.00   

       margin_gross_pow_ele  margin_net_pow_ele  nb_prod_act  net_margin  \
0                     25.44               25.44            2      678.99   
1                     16.38               16.38            1       18.89   
2                     28.60               28.60            1        6.60   
3                     30.22               30.22            1       25.46   
4                     44.91               44.91            1       47.98   
...                     ...                 ...          ...         ...   
14601                 27.88               27.88            2      381.77   
14602                  0.00                0.00            1       90.34   
14603                 39.84               39.84            1       20.38   
14604                 13.08               13.08            1        0.96   
14605                 11.84               11.84            1       96.34   

      num_years_antig                         origin_up  pow_max  churn  
0                   3  lxidpiddsbxsbosboudacockeimpuepw   43.648      1  
1                   6  kamkkxfxxuwbdslkwifmmcsiusiuosws   13.800      0  
2                   6  kamkkxfxxuwbdslkwifmmcsiusiuosws   13.856      0  
3                   6  kamkkxfxxuwbdslkwi

# conclusion:In the datset we have 14606 Rows & 26 Columns

In [10]:
df1 = pd.read_csv("price_data (1).csv")
df1

id  price_date  price_off_peak_var  \
0       038af19179925da21a25619c5a24b745  2015-01-01            0.151367   
1       038af19179925da21a25619c5a24b745  2015-02-01            0.151367   
2       038af19179925da21a25619c5a24b745  2015-03-01            0.151367   
3       038af19179925da21a25619c5a24b745  2015-04-01            0.149626   
4       038af19179925da21a25619c5a24b745  2015-05-01            0.149626   
...                                  ...         ...                 ...   
192997  16f51cdc2baa19af0b940ee1b3dd17d5  2015-08-01            0.119916   
192998  16f51cdc2baa19af0b940ee1b3dd17d5  2015-09-01            0.119916   
192999  16f51cdc2baa19af0b940ee1b3dd17d5  2015-10-01            0.119916   
193000  16f51cdc2baa19af0b940ee1b3dd17d5  2015-11-01            0.119916   
193001  16f51cdc2baa19af0b940ee1b3dd17d5  2015-12-01            0.119916   

        price_peak_var  price_mid_peak_var  price_off_peak_fix  \
0             0.000000            0.000000           44.266931   
1             0.000000            0.000000           44.266931   
2             0.000000            0.000000           44.266931   
3             0.000000            0.000000           44.266931   
4             0.000000            0.000000           44.266931   
...                ...                 ...                 ...   
192997        0.102232            0.076257           40.728885   
192998        0.102232            0.076257           40.728885   
192999        0.102232            0.076257           40.728885   
193000        0.102232            0.076257           40.728885   
193001        0.102232            0.076257           40.728885   

        price_peak_fix  price_mid_peak_fix  
0              0.00000            0.000000  
1              0.00000            0.000000  
2              0.00000            0.000000  
3              0.00000            0.000000  
4              0.00000            0.000000  
...                ...                 ...  
192997        24.43733           16.291555  
192998        24.43733           16.291555  
192999        24.43733           16.291555  
193000        24.43733           16.291555  
193001        24.43733           16.291555  

[193002 rows x 8 columns]

# conclusion : In the Price datset we have 193002 Rows & 8 columns

In [32]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.cluster import KMeans
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import IsolationForest
from sklearn.metrics import accuracy_score, classification_report, mean_squared_error
import numpy as np

# Reloading client data to ensure it's defined
client_data = pd.read_csv('client_data (1).csv')
client_data = client_data.fillna(0)

# Feature Selection for Churn Prediction
churn_features = ['cons_12m', 'cons_last_month', 'forecast_cons_12m', 'margin_gross_pow_ele', 'net_margin', 'pow_max']
X = client_data[churn_features]
y = client_data['churn']

# Split Data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Model for Churn Prediction
churn_model = RandomForestClassifier()
churn_model.fit(X_train, y_train)
y_pred = churn_model.predict(X_test)
churn_accuracy = accuracy_score(y_test, y_pred)

# Customer Segmentation using KMeans
kmeans = KMeans(n_clusters=3, random_state=42)
client_data['segment'] = kmeans.fit_predict(X)

# Revenue Forecasting (Net Margin)
X_revenue = client_data[['cons_12m', 'forecast_cons_12m', 'margin_gross_pow_ele', 'pow_max']]
y_revenue = client_data['net_margin']

X_train_rev, X_test_rev, y_train_rev, y_test_rev = train_test_split(X_revenue, y_revenue, test_size=0.2, random_state=42)
revenue_model = LinearRegression()
revenue_model.fit(X_train_rev, y_train_rev)
y_pred_rev = revenue_model.predict(X_test_rev)
revenue_mse = mean_squared_error(y_test_rev, y_pred_rev)

# Fraud Detection using Isolation Forest
fraud_model = IsolationForest(contamination=0.01, random_state=42)
client_data['fraud'] = fraud_model.fit_predict(X)

(churn_accuracy, revenue_mse, client_data[['segment', 'fraud']].head())


(0.8976728268309377,
 156433.14646972422,
    segment  fraud
 0        0      1
 1        0      1
 2        0      1
 3        0      1
 4        0      1)